### PR TITLE
DFR-3137: Remove user case access event

### PR DESCRIPTION
**Pull request by jthmcts**
### Jira link

https://tools.hmcts.net/jira/browse/DFR-3137

### Change description

Add event "Remove user case access".
This event is for both contested and consented case types.
Its purpose is to remove solicitor access from the case where this process has failed due to Notice of Change bugs.
The event is for superusers only.

### CODE REVIEW

**Review by al-hmcts — DISMISSED** [review]

**Review by hamzahtahirhmcts — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/removeusercaseaccess/RemoveUserCaseAccessAboutToStartHandler.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+    @Override
+    public GenericAboutToStartOrSubmitCallbackResponse<FinremCaseData> handle(
+        FinremCallbackRequest finremCallbackRequest, String authToken) {
+
+        String caseId = String.valueOf(finremCallbackRequest.getCaseDetails().getId());
+        log.info("Remove User Case Access about to start callback for Case ID: {}", caseId);
+
+        List<CaseAssignedUserRole> caseAssignedUserRoles = getCaseAssignedUserRoles(authToken, caseId);
+        List<DynamicListElement> dynamicListElements = caseAssignedUserRoles.stream()
```
Comment (hamzahtahirhmcts) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/removeusercaseaccess/RemoveUserCaseAccessAboutToStartHandler.java:59]: Not sure if it's necessary but could we not put this into a separate method too? 

Something like:


`getUserCaseAccessList(authToken, caseId)`

Then we can call `getCaseAssignedUserRoles` within this method.

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/removeusercaseaccess/RemoveUserCaseAccessAboutToStartHandler.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+        return caseAssignedUserRolesResource.getCaseAssignedUserRoles();
+    }
+
+    private UserCaseAccess getUserCaseAccess(String authToken, CaseAssignedUserRole caseAssignedUserRole) {
+        UserDetails userDetails = idamAuthApi.getUserByUserId(authToken, caseAssignedUserRole.getUserId());
+        return UserCaseAccess.of(caseAssignedUserRole, userDetails);
+    }
+
+    private DynamicListElement map(UserCaseAccess userCaseAccess) {
```
Comment (hamzahtahirhmcts) [inline src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/removeusercaseaccess/RemoveUserCaseAccessAboutToStartHandler.java:88]: Maybe the name should be more specific, something like `mapToDynamicListElement`.

**Review by hamzahtahirhmcts — DISMISSED** [review]

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/removeusercaseaccess/RemoveUserCaseAccessAboutToStartHandler.java (hunk 1 above)
Comment (jthmcts) [reply src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/removeusercaseaccess/RemoveUserCaseAccessAboutToStartHandler.java]: Method refactored

File: src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/removeusercaseaccess/RemoveUserCaseAccessAboutToStartHandler.java (hunk 2 above)
Comment (jthmcts) [reply src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/handler/removeusercaseaccess/RemoveUserCaseAccessAboutToStartHandler.java]: Method renamed

**Review by hamzahtahirhmcts — APPROVED** [review]

**Review by al-hmcts — APPROVED** [review]

**Pull request merged**